test/api/api_test.go

```diff
@@ -138,6 +139,72 @@ func TestWithDBSession(t *testing.T) {
 	}
 }
 
+// TestWithDBSession_RetryOnWriteConflict ensures that WithDBSession can
+// properly retry requests on MongoDB WriteConflict error.
+func TestWithDBSession_RetryOnWriteConflict(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	dbName := test.DBNameForTest(t.Name())
+	at, err := test.NewAccountsTester(dbName)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if errClose := at.Close(); errClose != nil {
+			t.Error(errors.AddContext(errClose, "failed to close account tester"))
+		}
+	}()
+
+	// Ensure WithDBSession works with requests without bodies.
+	// This is a regression test. It panics with a nil pointer if we cannot
+	// properly handle requests with nil bodies.
+	testAPI, err := api.New(at.DB, nil, at.Logger, nil)
+	if err != nil {
+		t.Fatal("Failed to instantiate API.", err)
+	}
+	handler := func(_ http.ResponseWriter, _ *http.Request, _ httprouter.Params) {}
+	req := (&http.Request{}).WithContext(context.Background())
+	testAPI.WithDBSession(handler)(nil, req, nil)
+
+	// Create a test user.
+	userEmail := types.NewEmail(t.Name() + "@siasky.net")
+	userPassword := t.Name() + "pass"
+	_, b, err := at.UserPOST(userEmail.String(), userPassword)
+	if err != nil {
+		t.Fatal(err, string(b))
+	}
+	defer func() {
+		_, _ = at.UserDELETE()
+	}()
+	r, b, err := at.LoginCredentialsPOST(userEmail.String(), userPassword)
+	if err != nil {
+		t.Fatal(err, string(b))
+	}
+	at.SetCookie(test.ExtractCookie(r))
+
+	// Set up several goroutines that will update the user simultaneously.
+	// We'll make them all block on a channel and then we'll close the channel,
+	// so they all start at the same time. We want to keep the number of
+	// conflicting goroutines low because we want the WriteConflict to resolve
+	// within the given dxTxnRetryCount attempts.
+	ch := make(chan int)
+	var wg sync.WaitGroup
+	for i := 0; i < 3; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			<-ch
```

Comment:
I'm not sure if this is sufficient to guarantee there is a write conflict, unless you wrote this test first and proved that current state had write conflicts with this test.

One way around this would be with a test dependency that randomly sleeps for a certain amount of time in the actual mongo write, so some txns will start, and then sleep for a second or two before finishing while others complete with no sleep.

Reply:
50 out of 50 is pretty good 😄 

Added comment on the dependency as to clear up what my original suggestion was.